File: lapras-agents/src/main/java/kr/ac/kaist/cdsn/lapras/agents/monnitserver/MonnitServerAgent.java
package kr.ac.kaist.cdsn.lapras.agents.monnitserver;

import com.monnit.mine.BaseApplication.BitConverter;
import com.monnit.mine.BaseApplication.Datum;
import com.monnit.mine.BaseApplication.applicationclasses.PIRBase;
import com.monnit.mine.BaseApplication.applicationclasses.TiltBase;
import com.monnit.mine.MonnitMineAPI.*;
import com.monnit.mine.MonnitMineAPI.enums.eFirmwareGeneration;
import com.monnit.mine.MonnitMineAPI.enums.eGatewayType;
import com.monnit.mine.MonnitMineAPI.enums.eMineListenerProtocol;
import com.monnit.mine.MonnitMineAPI.enums.eSensorApplication;
import kr.ac.kaist.cdsn.lapras.Agent;
import kr.ac.kaist.cdsn.lapras.AgentConfig;
import kr.ac.kaist.cdsn.lapras.agent.AgentComponent;
import kr.ac.kaist.cdsn.lapras.context.ContextManager;
import kr.ac.kaist.cdsn.lapras.event.EventDispatcher;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import java.nio.ByteBuffer;

import java.io.IOException;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.*;

/**
 * Created by JWP on 2017. 8. 1.
 *
 * MonnitServerAgent handles all sensors developed by Monnit.
 * Currently, it handles seat occupancy sensor and IR motion sensor, and door sensor.
 */

public class MonnitServerAgent extends AgentComponent {
    private static final Logger LOGGER = LoggerFactory.getLogger(MonnitServerAgent.class);

    private static final Map<Integer, String> seatSerialMap = new HashMap<>();
    private static final Map<Integer, String> motionSerialMap = new HashMap<>();
    private static final Map<Integer, String> doorSerialMap = new HashMap<>();
    private static final Map<Integer, String> tiltSerialMap = new HashMap<>(); // This is what I handle 4/27/2018-Kingberly

    private static final Map<String, Boolean> seatCountMap = new HashMap<>(); // map for total seat count

    private final List<Sensor> seatSensors = new ArrayList<>();
    private final List<Sensor> motionSensors = new ArrayList<>();
    private final List<Sensor> doorSensors = new ArrayList<>();
    private final List<Sensor> tiltSensors = new ArrayList<>(); // This is what I handle 4/27/2018-Kingberly

    private AgentConfig agentConfig;
    private ContextManager contextManager;
    private MineServer mineServer;

    /**
     * Initialize member variabless with the given configuration file
     */

    public MonnitServerAgent(EventDispatcher eventDispatcher, Agent agent) {
        super(eventDispatcher, agent);

        agentConfig = agent.getAgentConfig();
        contextManager = agent.getContextManager();

        String[] seatSensorIds = agentConfig.getOptionAsArray("seat_sensor_ids");
        String[] seatSensorNames = agentConfig.getOptionAsArray("seat_sensor_names");
        String[] motionSensorIds = agentConfig.getOptionAsArray("motion_sensor_ids");
        String[] motionSensorNames = agentConfig.getOptionAsArray("motion_sensor_names");
        String[] doorSensorIds = agentConfig.getOptionAsArray("door_sensor_ids");
        String[] doorSensorNames = agentConfig.getOptionAsArray("door_sensor_names");
        String[] tiltSensorIds = agentConfig.getOptionAsArray("tilt_sensor_ids");
        String[] tiltSensorNames = agentConfig.getOptionAsArray("tilt_sensor_names");

        initializeSeatCount();

        initializeSensors(seatSensorIds, seatSensorNames, seatSensors, seatSerialMap, eSensorApplication.Seat_Occupancy);
        initializeSensors(motionSensorIds, motionSensorNames, motionSensors, motionSerialMap, eSensorApplication.Infared_Motion);
        initializeSensors(doorSensorIds, doorSensorNames, doorSensors, doorSerialMap, eSensorApplication.Open_Closed);
        initializeSensors(tiltSensorIds, tiltSensorNames, tiltSensors, tiltSerialMap, eSensorApplication.Accelerometer_Tilt); //There are som addition for Pitch, Tilt
    }

    /**
     * Setup Monnit Mineserver
     * - register gateway to server
     * - register sensor to gateway
     * - attach sensor message handler to server
     */

    @Override
    public void setUp() {
        super.setUp();

        eMineListenerProtocol Protocol = eMineListenerProtocol.TCP;
        int port = Integer.parseInt(agentConfig.getOption("mineserver_port"));
        int gatewayID = Integer.parseInt(agentConfig.getOption("gateway_id"));
        String ipaddr = agentConfig.getOption("mineserver_ip");

        InetAddress ip;

        try {
            ip = InetAddress.getByName(ipaddr);
            mineServer = new MineServer(Protocol, ip, port);
            mineServer.StartServer();

            LOGGER.info("MineServer Constructed");

            //Add handler
            if (mineServer.addSensorDataProcessingHandler(new SensorMessageHandler())) {
                LOGGER.info("Successfully added SensorHandler");
            } else {
                LOGGER.info("Failed to add handler.");
            }

            //Register Gateway
            Gateway usbGate = new Gateway(gatewayID, eGatewayType.USBService);
            usbGate.UpdateReportInterval(1);
            mineServer.RegisterGateway(usbGate);

            //Register sensors
            for (Sensor seatSensor : seatSensors) {
                mineServer.RegisterSensor(usbGate.GatewayID, seatSensor);
            }

            for (Sensor motionSensor : motionSensors) {
                mineServer.RegisterSensor(usbGate.GatewayID, motionSensor);
                PIRBase.SensorEdit(motionSensor, null, 0.1, 2, 1, null);
            }

            for (Sensor doorSensor : doorSensors) {
                mineServer.RegisterSensor(usbGate.GatewayID, doorSensor);
            }

            for (Sensor tiltSensor : tiltSensors) {
                mineServer.RegisterSensor(usbGate.GatewayID, tiltSensor);
                TiltBase.UpdateSensorHeartBeat(tiltSensor, 1, true); //To-do: slowly?
            }
        } catch (UnknownHostException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        } catch (Exception e) {
            e.printStackTrace();
        }

    }

    @Override
    public void run() {

        while (true) {
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                LOGGER.debug("Stop Server!");
                try {
                    mineServer.StopServer();
                } catch (InterruptedException e1) {
                    e1.printStackTrace();
                } catch (IOException e1) {
                    e1.printStackTrace();
                }
                e.printStackTrace();
            }
        }
    }

    /**
     * Intialize sensors & contexts
     */

    private void initializeSensors (String[] sensorIds, String[] sensorNames, final List <Sensor> sensors,
                                   final Map<Integer, String> serialMap, eSensorApplication sensorType) {
        int index = 0;

        for (String sensorId : sensorIds) {
            int sensorIdNum = Integer.parseInt(sensorId);
            try {
                sensors.add(new Sensor(sensorIdNum, sensorType, "2.5.4.0", eFirmwareGeneration.Commercial));
                serialMap.put(sensorIdNum, sensorNames[index]);
                contextManager.updateContext(sensorNames[index], false, agentName);
                if(sensorNames[index].contains("tilt")){  //Tilt needs two component. 'Pitch' & 'Roll'
                    contextManager.setPublishAsUpdated(String.format("%s_%s",sensorNames[index],"Pitch"));
                    contextManager.setPublishAsUpdated(String.format("%s_%s",sensorNames[index],"Roll"));
                }else {
                    contextManager.setPublishAsUpdated(sensorNames[index]);
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
            index ++;
        }

    }

    /**
     * Initalize seat count
     */

    private void initializeSeatCount() {
        String[] seatSensorNames = agentConfig.getOptionAsArray("seat_sensor_names");

        for (String seatSensorName : seatSensorNames) {
            seatCountMap.put(seatSensorName, false);
        }

        contextManager.updateContext("totalSeatCount", 0, agentName);
        contextManager.setPublishAsUpdated("totalSeatCount");
    }

    /**
     * Class for parsing sensor messages & publishing the appropriate context message.
     */

    private class SensorMessageHandler implements iSensorMessageHandler {

        @Override
        public void ProcessSensorMessages(List<SensorMessage> sensorMessageList, Gateway gateway) throws Exception {

            for (SensorMessage msg : sensorMessageList) {
                LOGGER.info(msg.toString());

                if (seatSerialMap.get(msg.SensorID) != null) {
                    handleSeatSensorMessage(msg);
                } else if (motionSerialMap.get(msg.SensorID) != null) {
                    handleMotionSensorMessage(msg);
                } else if (doorSerialMap.get(msg.SensorID) != null) {
                    handleDoorSensorMessage(msg);
                } else if (tiltSerialMap.get(msg.SensorID) != null) {
                    handleTiltSensorMessage(msg);

                } else {
                    LOGGER.info("Sensor is not registered");
                }
            }
        }

        private void handleSeatSensorMessage(SensorMessage msg) {
            String contextName = seatSerialMap.get(msg.SensorID);
            Long messageTime = msg.getMessageDate().getTimeInMillis();

            if (contextName != null) {
                LOGGER.info("{} updated", contextName);
                contextManager.updateContext(contextName, msg.State == 2, agentName, messageTime);
                seatCountMap.put(contextName, msg.State == 2);
            }

            int totalSeatCount = Collections.frequency(seatCountMap.values(), true);
            LOGGER.debug("seat count: " + seatCountMap);
            contextManager.updateContext("totalSeatCount", totalSeatCount, agentName, messageTime);
        }

        private void handleMotionSensorMessage(SensorMessage msg) {
            String contextName = motionSerialMap.get(msg.SensorID);
            Long messageTime = msg.getMessageDate().getTimeInMillis();

            if (contextName != null) {
                LOGGER.info("{} updated", contextName);
                contextManager.updateContext(contextName, msg.State == 2, agentName, messageTime);
            }
        }

        private void handleDoorSensorMessage(SensorMessage msg) {
            String contextName = doorSerialMap.get(msg.SensorID);
            Long messageTime = msg.getMessageDate().getTimeInMillis();

            if (contextName != null) {
                LOGGER.info("{} updated", contextName);
                contextManager.updateContext(contextName, msg.State == 2, agentName, messageTime);
            }
        }
        public byte[] swapByte(byte[] data) {
            byte temp;
            temp = data[1];
            data[1] = data[0];
            data[0] = temp;
            return data;
        }
        public short convertToShort(byte[] array) {
            ByteBuffer buffer = ByteBuffer.wrap(array);
            return buffer.getShort();
        }

        private void handleTiltSensorMessage(SensorMessage msg) throws Exception {
            String contextName = tiltSerialMap.get(msg.SensorID);
            Long messageTime = msg.getMessageDate().getTimeInMillis();

            if (contextName != null) {
                LOGGER.info("{} updated", contextName);

                List<Datum> data = msg.getData(); // Pitch, Roll value
                for (Datum element : data) {
                    String description = element.Description;

                    //Change Network bit to Host bit (Big-Endian)
                    Double doubleData = (double)element.Data*100;  //consider second decimal place.
                    short intData = doubleData.shortValue();    //convert to short
                    byte[] byteData = BitConverter.getBytes(intData); //read intData to byteData
                    byteData = swapByte(byteData);  //byte swap!(NetworkBit to HostBit)
                    short temp = convertToShort(byteData); //convert to short
                    Object value = (double) temp/100; //convert to double

                    //To-do: how to publish context value?
                    contextManager.updateContext(String.format("%s_%s", contextName, description), value, agentName, messageTime);
                }
            }
        }
    }
}

File: lapras_agents/monnit_server_agent.py
import logging
from typing import Dict, List, Set
from dataclasses import dataclass
from datetime import datetime
import time
from lapras_middleware.agent import AgentComponent
from lapras_middleware.context import Context, ContextField
from lapras_middleware.event import Event, EventDispatcher

logger = logging.getLogger(__name__)

class MonnitServerAgent(AgentComponent):
    """
    MonnitServerAgent handles all sensors developed by Monnit.
    Currently handles seat occupancy sensor, IR motion sensor, and door sensor.
    """
    
    def __init__(self, event_dispatcher: EventDispatcher, agent):
        super().__init__(event_dispatcher, agent)
        
        # Initialize maps for different sensor types
        self.seat_serial_map: Dict[int, str] = {}
        self.motion_serial_map: Dict[int, str] = {}
        self.door_serial_map: Dict[int, str] = {}
        self.tilt_serial_map: Dict[int, str] = {}
        
        # Map for total seat count
        self.seat_count_map: Dict[str, bool] = {}
        
        # Lists for different sensor types
        self.seat_sensors: List = []
        self.motion_sensors: List = []
        self.door_sensors: List = []
        self.tilt_sensors: List = []
        
        # Get configuration
        self.agent_config = agent.agent_config
        self.context_manager = agent.context_manager
        
        # Initialize sensors from config
        self._initialize_seat_count()
        self._initialize_sensors()
        
    def _initialize_seat_count(self):
        """Initialize seat count from configuration"""
        seat_sensor_names = self.agent_config.get_option_as_array("seat_sensor_names")
        
        for seat_name in seat_sensor_names:
            self.seat_count_map[seat_name] = False
            
        self.context_manager.update_context("totalSeatCount", 0, self.agent_name)
        self.context_manager.set_publish_as_updated("totalSeatCount")
        
    def _initialize_sensors(self):
        """Initialize all sensor types from configuration"""
        # Initialize seat sensors
        seat_sensor_ids = self.agent_config.get_option_as_array("seat_sensor_ids")
        seat_sensor_names = self.agent_config.get_option_as_array("seat_sensor_names")
        self._init_sensor_type(seat_sensor_ids, seat_sensor_names, 
                             self.seat_sensors, self.seat_serial_map, "Seat_Occupancy")
        
        # Initialize motion sensors
        motion_sensor_ids = self.agent_config.get_option_as_array("motion_sensor_ids")
        motion_sensor_names = self.agent_config.get_option_as_array("motion_sensor_names")
        self._init_sensor_type(motion_sensor_ids, motion_sensor_names,
                             self.motion_sensors, self.motion_serial_map, "Infared_Motion")
        
        # Initialize door sensors
        door_sensor_ids = self.agent_config.get_option_as_array("door_sensor_ids")
        door_sensor_names = self.agent_config.get_option_as_array("door_sensor_names")
        self._init_sensor_type(door_sensor_ids, door_sensor_names,
                             self.door_sensors, self.door_serial_map, "Open_Closed")
        
        # Initialize tilt sensors
        tilt_sensor_ids = self.agent_config.get_option_as_array("tilt_sensor_ids")
        tilt_sensor_names = self.agent_config.get_option_as_array("tilt_sensor_names")
        self._init_sensor_type(tilt_sensor_ids, tilt_sensor_names,
                             self.tilt_sensors, self.tilt_serial_map, "Accelerometer_Tilt")
                             
    def _init_sensor_type(self, sensor_ids: List[str], sensor_names: List[str],
                         sensors: List, serial_map: Dict[int, str], sensor_type: str):
        """Initialize a specific type of sensor"""
        for idx, sensor_id in enumerate(sensor_ids):
            try:
                sensor_id_num = int(sensor_id)
                # Create sensor object (implementation depends on actual sensor class)
                sensor = self._create_sensor(sensor_id_num, sensor_type)
                sensors.append(sensor)
                serial_map[sensor_id_num] = sensor_names[idx]
                
                # Initialize context
                self.context_manager.update_context(sensor_names[idx], False, self.agent_name)
                
                # Handle tilt sensors specially
                if "tilt" in sensor_names[idx].lower():
                    self.context_manager.set_publish_as_updated(f"{sensor_names[idx]}_Pitch")
                    self.context_manager.set_publish_as_updated(f"{sensor_names[idx]}_Roll")
                else:
                    self.context_manager.set_publish_as_updated(sensor_names[idx])
                    
            except Exception as e:
                logger.error(f"Error initializing sensor {sensor_id}: {e}")
                
    def _create_sensor(self, sensor_id: int, sensor_type: str):
        """Create a sensor object based on type"""
        # This is a placeholder - actual implementation would depend on sensor hardware
        return {
            'id': sensor_id,
            'type': sensor_type,
            'version': '2.5.4.0',
            'generation': 'Commercial'
        }
        
    def set_up(self):
        """Set up the Monnit server and register sensors"""
        super().set_up()
        
        try:
            # Get configuration
            protocol = "TCP"
            port = int(self.agent_config.get_option("mineserver_port"))
            gateway_id = int(self.agent_config.get_option("gateway_id"))
            ip_addr = self.agent_config.get_option("mineserver_ip")
            
            # Initialize server
            self._init_server(protocol, ip_addr, port, gateway_id)
            
        except Exception as e:
            logger.error(f"Error setting up Monnit server: {e}")
            
    def _init_server(self, protocol: str, ip_addr: str, port: int, gateway_id: int):
        """Initialize the Monnit server"""
        # This is a placeholder - actual implementation would depend on Monnit API
        logger.info("Initializing Monnit server")
        # Here you would:
        # 1. Create server connection
        # 2. Register gateway
        # 3. Register sensors
        # 4. Set up message handlers
        
    def run(self):
        """Main run loop"""
        while True:
            try:
                time.sleep(1)
            except InterruptedException as e:
                logger.debug("Stopping server!")
                self._stop_server()
                break
                
    def _stop_server(self):
        """Stop the Monnit server"""
        try:
            # Actual server shutdown implementation would go here
            logger.info("Server stopped")
        except Exception as e:
            logger.error(f"Error stopping server: {e}")
            
    def handle_sensor_message(self, message: dict):
        """Handle incoming sensor messages"""
        try:
            sensor_id = message.get('sensor_id')
            state = message.get('state')
            timestamp = message.get('timestamp', int(time.time() * 1000))
            
            if sensor_id in self.seat_serial_map:
                self._handle_seat_sensor_message(sensor_id, state, timestamp)
            elif sensor_id in self.motion_serial_map:
                self._handle_motion_sensor_message(sensor_id, state, timestamp)
            elif sensor_id in self.door_serial_map:
                self._handle_door_sensor_message(sensor_id, state, timestamp)
            elif sensor_id in self.tilt_serial_map:
                self._handle_tilt_sensor_message(sensor_id, state, timestamp)
            else:
                logger.info("Sensor is not registered")
                
        except Exception as e:
            logger.error(f"Error handling sensor message: {e}")
            
    def _handle_seat_sensor_message(self, sensor_id: int, state: int, timestamp: int):
        """Handle seat sensor messages"""
        context_name = self.seat_serial_map.get(sensor_id)
        if context_name:
            logger.info(f"{context_name} updated")
            is_occupied = (state == 2)
            self.context_manager.update_context(context_name, is_occupied, self.agent_name, timestamp)
            self.seat_count_map[context_name] = is_occupied
            
            # Update total seat count
            total_occupied = sum(1 for occupied in self.seat_count_map.values() if occupied)
            self.context_manager.update_context("totalSeatCount", total_occupied, self.agent_name, timestamp)
            
    def _handle_motion_sensor_message(self, sensor_id: int, state: int, timestamp: int):
        """Handle motion sensor messages"""
        context_name = self.motion_serial_map.get(sensor_id)
        if context_name:
            logger.info(f"{context_name} updated")
            self.context_manager.update_context(context_name, state == 2, self.agent_name, timestamp)
            
    def _handle_door_sensor_message(self, sensor_id: int, state: int, timestamp: int):
        """Handle door sensor messages"""
        context_name = self.door_serial_map.get(sensor_id)
        if context_name:
            logger.info(f"{context_name} updated")
            self.context_manager.update_context(context_name, state == 2, self.agent_name, timestamp)
            
    def _handle_tilt_sensor_message(self, sensor_id: int, state: int, timestamp: int):
        """Handle tilt sensor messages"""
        context_name = self.tilt_serial_map.get(sensor_id)
        if context_name:
            logger.info(f"{context_name} updated")
            # Implementation would depend on tilt sensor data format
            # You might need to handle pitch and roll separately 